File: solutions/swift/robot-simulator/3/Sources/RobotSimulator/RobotSimulator.swift
enum Direction: Int { case north, east, south, west }

struct RobotState {
  var x, y: Int
  var bearing: Direction

  mutating func turnLeft() { bearing = Direction(rawValue: (bearing.rawValue + 3) % 4)! }
  mutating func turnRight() { bearing = Direction(rawValue: (bearing.rawValue + 1) % 4)! }
  mutating func advance() {
    (x, y) = (
      x + (bearing == .east ? 1 : bearing == .west ? -1 : 0),
      y + (bearing == .north ? 1 : bearing == .south ? -1 : 0)
    )
  }
}

class SimulatedRobot {
  var state: RobotState

  init(x: Int, y: Int, bearing: Direction) {
    state = RobotState(x: x, y: y, bearing: bearing)
  }

  func move(commands: String) {
    for c in commands {
      switch c {
      case "L": state.turnLeft()
      case "R": state.turnRight()
      case "A": state.advance()
      default: fatalError()
      }
    }
  }
}

File: solutions/lua/robot-simulator/2/robot-simulator.lua
local R = { north = "east", east = "south", south = "west", west = "north" }
local L = { north = "west", east = "north", south = "east", west = "south" }
local P = { north = { 0, 1 }, east = { 1, 0 }, south = { 0, -1 }, west = { -1, 0 } }
return function(config)
  config.move = function(self, commands)
    commands:gsub('.', function(c)
      if c == 'A' then self.x, self.y = self.x + P[self.heading][1], self.y + P[self.heading][2]
      elseif c == 'L' then self.heading = L[self.heading]
      elseif c == 'R' then self.heading = R[self.heading]
      else error("Unknown command: " .. c)
      end
    end)
  end
  return config
end
